Alerting:  New alert list filter improvements

konrad147 opened the pull request:
**What is this feature?**
This PR adds a few improvements to the filters on the new list page. It also improves filtering performance
The PR also adds a new `queriedDatasourcesUIDs` field to GMA prometheus API response to make it easy to filter rules by used data sources based on Prom API

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

gillesdemey committed:
search WIP fix

konrad147 committed:
Move filtering code to generators for performance reasons

Discarding rules and groups early in the iterable chain limits the number of promises we need to wait for which improves performance significantly

konrad147 committed:
Add error handling for generators

konrad147 committed:
Add support for data source filter for GMA rules

konrad147 committed:
Fix datasource filter

konrad147 committed:
Move filtering back to filtered rules hook, use paged groups for improved performance

konrad147 committed:
Add queriedDatasources field to grafana managed rules and update filtering logic to rely on it

- Introduced a new field `queriedDatasources` in the AlertingRule struct to track data sources used in rules.
- Updated the Prometheus API to populate `queriedDatasources` when creating alerting rules.
- Modified filtering logic in the ruleFilter function to utilize the new `queriedDatasources` field for improved data source matching.
- Adjusted related tests to reflect changes in rule structure and filtering behavior.

moustafab commented on pkg/services/ngalert/api/tooling/definitions/prom.go:
```diff
@@ -168,12 +169,10 @@ type AlertingRule struct {
 	// required: true
-	UID string `json:"uid"`
-	// required: true
-	Name string `json:"name"`
-	// required: true
-	FolderUID string `json:"folderUid"`
+	Name      string `json:"name"`
+	FolderUID string `json:"folderUid,omitempty"`
```
was the removal of the `required: true` intentional?

konrad147 replied:
Yes. Unfortunatelly, the `Rule` struct seems to be shared between GMA and DMA rules, and with `required: true` the API returns empty string for DMA rules for `FolderUID` and `UID` properties
If there is any better way to fix this I'm happy to do that

moustafab commented on pkg/services/ngalert/api/tooling/definitions/prom.go:
```diff
@@ -152,9 +152,10 @@ type AlertingRule struct {
-	Duration      float64 `json:"duration,omitempty"`
-	KeepFiringFor float64 `json:"keepFiringFor,omitempty"`
+	Query              string   `json:"query,omitempty"`
+	QueriedDatasources []string `json:"queriedDatasources,omitempty"`
```
Nit: maybe a bit clearer what the string actually is?
```suggestion
	QueriedDatasourceUIDs []string `json:"queriedDatasourceUIDs,omitempty"`
```

konrad147 committed:
Add FilterView performance logging

konrad147 committed:
Improve GMA Prometheus types, rename queried datasources property

konrad147 committed:
Use custom generator helpers for flattening and filtering rule groups

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/list-v2-filter-improvements

konrad147 committed:
Fix lint errors, add missing translations

konrad147 committed:
Revert test condition

konrad147 committed:
Refactor api prom changes

konrad147 committed:
Fix lint errors

konrad147 committed:
Update backend tests

tomratcliffe commented on public/app/features/alerting/unified/rule-list/FilterView.test.tsx:
```diff
@@ -40,9 +40,6 @@ beforeEach(() => {
 const io = mockIntersectionObserver();
 
 describe('RuleList - FilterView', () => {
-  jest.setTimeout(60 * 1000);
-  jest.retryTimes(2);
```
(I'm all for removing if unneeded, but just asking) Has something changed in the implementation that means these are more reliable now?

konrad147 replied:
Yes, we’ve changed how filters are applied to generators, which has significantly improved performance for the FilterView, making this test run much faster.

tomratcliffe commented on public/app/features/alerting/unified/rule-list/GrafanaRuleLoader.tsx:
```diff
@@ -38,7 +44,22 @@ export function GrafanaRuleLoader({ rule, groupIdentifier, namespaceName }: Graf
     }
 
-    return <AlertRuleListItemSkeleton />;
+    if (isLoading) {
```
Could this and the isError check be moved up above the `if (!rulerRule)` check?

tomratcliffe commented on public/app/types/unified-alerting-dto.ts:
```diff
@@ -122,7 +122,13 @@ interface PromRuleDTOBase {
   lastEvaluation?: string;
   lastError?: string;
-  uid?: string;
+  // uid?: string;
```
Commented out here?

konrad147 replied:
To be removed

gillesdemey commented on public/app/features/alerting/unified/rule-list/hooks/prometheusGroupsGenerator.ts:
```diff
@@ -95,6 +96,23 @@ export function useGrafanaGroupsGenerator(hookOptions: UseGeneratorHookOptions =
+ * @param generator - The paginated generator to convert
+ * @returns A non-paginated generator that yields all groups from the original generator one by one
+ */
+export function toPageless<TGroup extends PromRuleGroupDTO>(
```
nit: I was initially a bit confused about what this function did from the consuming side; perhaps something like `yieldIndividualRuleGroups` would be a bit more verbose but clearer when reading?

gillesdemey commented on pkg/services/ngalert/api/tooling/definitions/prom.go:
```diff
@@ -152,9 +152,10 @@ type AlertingRule struct {
-	Duration      float64 `json:"duration,omitempty"`
-	KeepFiringFor float64 `json:"keepFiringFor,omitempty"`
+	Query                 string   `json:"query,omitempty"`
+	QueriedDatasourceUIDs []string `json:"queriedDatasourceUIDs,omitempty"`
```
not entirely sure if this has been bike shed yet; but it took me a while to figure out that this is where we store the list of data query IDs that appear in the alert rule definition (initially I thought this was something that was computed at run-time when the alert rule had been evaluated).

can we try and find a clearer name?

gillesdemey replied:
maybe `DataSourceUIDsFromRuleDefinition`? 😬

konrad147 replied:
Wouldn't it be confusing in the other direction? Most of the fields in this struct come from the rule definition, but only one would be called this way

gillesdemey reviewed (commented):
releasing a few comments, still have a few files left to review :)

gillesdemey commented on public/app/features/alerting/unified/rule-list/hooks/useFilteredRulesIterator.ts:
```diff
@@ -52,48 +48,113 @@ export function useFilteredRulesIteratorProvider() {
+}
+
+// find all data sources that the user might want to filter by, only allow Prometheus and Loki data source types
+const SUPPORTED_RULES_SOURCE_TYPES = ['loki', 'prometheus'];
```
nit: hoist this somewhere in a constants file maybe?

konrad147 replied:
Fortunately, we already have such array defined in a different file 😃

konrad147 committed:
Refactor rule list components to improve error handling and data source management

- Enhanced error handling in FilterViewResults by logging errors before returning an empty iterable.
- Simplified conditional rendering in GrafanaRuleLoader for better readability.
- Updated data source handling in PaginatedDataSourceLoader and PaginatedGrafanaLoader to use new individual rule group generator.
- Renamed toPageless function to toIndividualRuleGroups for clarity in prometheusGroupsGenerator.
- Improved filtering logic in useFilteredRulesIterator to utilize a dedicated function for data source type validation.
- Added isRulesDataSourceType utility function for better data source type checks.
- Removed commented-out code in PromRuleDTOBase for cleaner interface definition.

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/list-v2-filter-improvements

gillesdemey commented:
before merging this PR; I've noticed the signal from the abort controller is either not propagating properly or something else is not correctly interrupting the AsyncIterable 🤔

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/list-v2-filter-improvements

konrad147 committed:
Fix abort controller on FilterView

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/list-v2-filter-improvements

konrad147 committed:
Improve generators filtering

gillesdemey committed:
fix abort controller

gillesdemey committed:
refactor cancelSearch

gillesdemey committed:
make states exclusive

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/list-v2-filter-improvements

konrad147 committed:
Load full page in one loadResultPage call

konrad147 committed:
Update tests, update translations

konrad147 commented:
/deploy-to-hg

gillesdemey committed:
Refactor filter status into separate component

gillesdemey committed:
hoist hook

gillesdemey reviewed (approved):
Amazing work on this one – thanks for sticking with all of the changes requested and persevering until we've got the iterables / cancellation all figured out and working properly ♥️

Let's get this shipped! :shipit:

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/list-v2-filter-improvements

tomratcliffe reviewed (approved):
:rocket: :shipit: Great job!

fayzal-g reviewed (approved):
Happy with it on the backend side

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/list-v2-filter-improvements

konrad147 committed:
Use the new function for supported rules source type
